File: sdk-go/example/announce_demo.go
// Conformance demo: a tiny net/http server, announced to the local agent.
// Run a local agent (./gruve), then, from sdk-go/:  go run ./example
package main

import (
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"syscall"

	gruve "github.com/MLTQ/gruve-kit/sdk-go"
)

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		fmt.Fprint(w, "<body style='background:#141821;color:#8ce0b0;font-family:monospace;"+
			"display:grid;place-items:center;height:100vh'><h1>hello from go \U0001f439</h1></body>")
	})

	// 1.4: listen BEFORE announcing — the agent probes the port.
	ln, err := net.Listen("tcp", "127.0.0.1:9703")
	if err != nil {
		panic(err)
	}
	fmt.Println("go demo app on :9703")
	go http.Serve(ln, mux)

	h, err := gruve.Announce(gruve.Options{
		ID: "godemo", Name: "Go Demo", Port: 9703,
		Blurb: "announced by gruve-sdk (go)", Hue: 140,
	})
	if err != nil {
		panic(err)
	}
	fmt.Println("announced; check the lobby. ctrl-c to exit (withdraws).")

	// Signal handling belongs to the app, not the SDK: withdraw on Ctrl-C.
	sig := make(chan os.Signal, 1)
	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
	<-sig
	h.Stop()
}

File: sdk-py/examples/announce_demo.py
"""Conformance demo: a tiny stdlib HTTP server, announced to the local agent.

Run a local agent (`./gruve`), then:  python examples/announce_demo.py
(from the sdk-py/ directory, or after `pip install -e .`)
"""
from http.server import BaseHTTPRequestHandler, HTTPServer

import gruve_sdk


class Hello(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header("content-type", "text/html")
        self.end_headers()
        self.wfile.write(
            b"<body style='background:#141821;color:#7aa2f7;font-family:monospace;"
            b"display:grid;place-items:center;height:100vh'>"
            b"<h1>hello from python \xf0\x9f\x90\x8d</h1></body>"
        )

    def log_message(self, *_):  # keep the console quiet
        pass


# 1.4: listen BEFORE announcing — the agent probes the port.
srv = HTTPServer(("127.0.0.1", 9702), Hello)
print("python demo app on :9702")

gruve = gruve_sdk.announce(
    id="pydemo", name="Python Demo", port=9702,
    blurb="announced by gruve-sdk (python)", hue=210,
)
print("announced; check the lobby. ctrl-c to exit (withdraws).")

try:
    srv.serve_forever()
except KeyboardInterrupt:
    gruve.stop()
